openrouteservice/src/main/java/org/heigit/ors/fastisochrones/partitioning/InertialFlow.java

```diff
@@ -0,0 +1,333 @@
+package org.heigit.ors.fastisochrones.partitioning;
+
+import com.carrotsearch.hppc.IntArrayList;
+import com.carrotsearch.hppc.IntHashSet;
+import com.carrotsearch.hppc.cursors.IntCursor;
+import com.graphhopper.routing.util.EdgeFilter;
+import com.graphhopper.storage.Graph;
+import com.graphhopper.storage.GraphHopperStorage;
+import com.graphhopper.util.EdgeExplorer;
+import com.graphhopper.util.EdgeIterator;
+import org.heigit.ors.fastisochrones.partitioning.Projector.Projection;
+import org.heigit.ors.routing.graphhopper.extensions.edgefilters.EdgeFilterSequence;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.concurrent.ExecutorService;
+
+import static org.heigit.ors.fastisochrones.partitioning.FastIsochroneParameters.*;
+
+/**
+ * Recursive implementation of InertialFlow algorithm for partitioning a graph.
+ *
+ * @author Hendrik Leuschner
+ */
+public class InertialFlow implements Runnable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(InertialFlow.class);
+    private static final int MIN_SPLITTING_ITERATION = 0;
+    private static final int MAX_SPLITTING_ITERATION = 268435456; //==2^28
+    private static final int MAX_SUBCELL_NUMBER = 10;
+    private static final boolean SEPARATEDISCONNECTED = true;
+    private static final int CONSIDERED_PROJECTIONS = 3;
+    private static Projector projector = new Projector();
+    protected Map<Projection, IntArrayList> projections;
+    private int cellId;
+    private Graph ghGraph;
+    private EdgeFilter edgeFilter;
+    private PartitioningData pData;
+    private int[] nodeToCellArr;
+    private ExecutorService executorService;
+    private InverseSemaphore inverseSemaphore;
+
+    public InertialFlow(int[] nodeToCellArray, GraphHopperStorage ghStorage, EdgeFilterSequence edgeFilters, ExecutorService executorService, InverseSemaphore inverseSemaphore) {
+        //Start cellId 1 so that bitshifting it causes no zeros at the front
+        setNodeToCellArr(nodeToCellArray);
+        setCellId(1);
+        setGraph(ghStorage.getBaseGraph());
+        setEdgeFilter(edgeFilters);
+        setExecutorService(executorService);
+        setInverseSemaphore(inverseSemaphore);
+
+        PartitioningData partitioningData = new PartitioningData();
+        PartitioningDataBuilder partitioningDataBuilder = new PartitioningDataBuilder(ghStorage.getBaseGraph(), partitioningData);
+        partitioningDataBuilder.setAdditionalEdgeFilter(edgeFilter);
+        partitioningDataBuilder.run();
+        setPartitioningData(partitioningData);
+        projector.setGHStorage(ghStorage);
+        setProjections(projector.calculateProjections());
+
+        LOGGER.info("Number of nodes: {}", ghGraph.getNodes());
+        LOGGER.info("Number of edges: {}", ghGraph.getAllEdges().length());
+    }
+
+    private InertialFlow() {
```

Comment:
Necessary?

Reply:
Do you mean the default constructor? Yes, it is used in the recursive creation of IntertialFlow, compared to the other constructor which is only called once for the very first iteration.

Reply:
Thanks for clarifying!